pkg/targetgroupbinding/targets_manager.go

```diff
@@ -0,0 +1,276 @@
+		targetsCacheItem.mutex.Lock()
+		defer targetsCacheItem.mutex.Unlock()
+		refreshedTargets, err := m.refreshOngoingTargets(ctx, tgARN, targetsCacheItem.targets)
+		if err != nil {
+			return nil, err
+		}
+		targetsCacheItem.targets = refreshedTargets
+		return cloneTargetInfoSlice(refreshedTargets), nil
+	}
+
+	refreshedTargets, err := m.refreshAllTargets(ctx, tgARN)
+	if err != nil {
+		return nil, err
+	}
+	targetsCacheItem := &targetsCacheItem{
+		mutex:   sync.RWMutex{},
+		targets: refreshedTargets,
+	}
+	m.targetsCache.Set(tgARN, targetsCacheItem, m.targetsCacheTTL)
+	return cloneTargetInfoSlice(refreshedTargets), nil
+}
+
+func (m *cachedTargetsManager) refreshAllTargets(ctx context.Context, tgARN string) ([]TargetInfo, error) {
+	targets, err := m.listTargetsFromAWS(ctx, tgARN, nil)
+	if err != nil {
+		return nil, err
+	}
+	return targets, nil
+}
+
+func (m *cachedTargetsManager) refreshOngoingTargets(ctx context.Context, tgARN string, cachedTargets []TargetInfo) ([]TargetInfo, error) {
+	var refreshedTargets []TargetInfo
+	var ongoingTargets []elbv2sdk.TargetDescription
+	for _, cachedTarget := range cachedTargets {
+		if cachedTarget.IsOngoing() {
+			ongoingTargets = append(ongoingTargets, cachedTarget.Target)
+		} else {
+			refreshedTargets = append(refreshedTargets, cachedTarget)
+		}
+	}
+	if len(ongoingTargets) == 0 {
+		return refreshedTargets, nil
+	}
+
+	refreshedOngoingTargets, err := m.listTargetsFromAWS(ctx, tgARN, ongoingTargets)
+	if err != nil {
+		return nil, err
+	}
+	for _, target := range refreshedOngoingTargets {
+		if target.IsNotRegistered() {
+			continue
+		}
+		refreshedTargets = append(refreshedTargets, target)
+	}
+	return refreshedTargets, nil
+}
+
+// listTargetsFromAWS will list targets for TargetGroup using ELBV2API.
+// if specified targets is non-empty, only these targets will be listed.
+// otherwise, all targets for targetGroup will be listed.
+func (m *cachedTargetsManager) listTargetsFromAWS(ctx context.Context, tgARN string, targets []elbv2sdk.TargetDescription) ([]TargetInfo, error) {
+	fmt.Printf("list %v targets\n", len(targets))
+	req := &elbv2sdk.DescribeTargetHealthInput{
+		TargetGroupArn: aws.String(tgARN),
+		Targets:        pointerizeTargetDescriptions(targets),
+	}
+	resp, err := m.elbv2Client.DescribeTargetHealthWithContext(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+
+	listedTargets := make([]TargetInfo, 0, len(resp.TargetHealthDescriptions))
+	for _, elem := range resp.TargetHealthDescriptions {
+		listedTargets = append(listedTargets, TargetInfo{
+			Target:       *elem.Target,
+			TargetHealth: elem.TargetHealth,
+		})
+	}
+	return listedTargets, nil
+}
+
+// recordSuccessfulDeregisterTargetsOperation will record a successful deregisterTarget operation
+func (m *cachedTargetsManager) recordSuccessfulRegisterTargetsOperation(tgARN string, targets []elbv2sdk.TargetDescription) {
+	rawTargetsCacheItem, exists := m.targetsCache.Get(tgARN)
```

Comment:
does it not need the targetCacheMutex?

Reply:
ideally, we don't need the targetCacheMutex here, since the cache access itself is already synchronized via mutex inside the cache implementation.
Also, for List/Register apicalls for the same tgARN won't happen concurrently, so we don't need to lock on targetCacheMutex here.

However, to make the interface itself secure without checking the implementation. I added a RLock here to make the Register/List for same tgARN safe in semantic too. (it hurts performance a little, ideally it should be a lock per targrtGroup).